File: leetcode/cpp/min-stack.cpp
#include <bits/stdc++.h>
using namespace std;

class MinStack
{
    // Naive Approach
public:
    stack<pair<int, int>> stk;
    MinStack()
    {
        ;
    }

    void push(int val)
    {
        if (stk.size() == 0)
        {
            stk.push({val, val});
        }
        else
        {
            int mini = min(val, stk.top().second);
            stk.push({val, mini});
        }
    }

    void pop()
    {
        if (stk.size() > 0)
        {
            stk.pop();
        }
    }

    int top()
    {
        return stk.top().first;
    }

    int getMin()
    {
        return stk.top().second;
    }
};

File: leetcode/python/min-stack.py
from typing import List
class MinStack:
    stack: List[tuple[int,int]] = []
    def __init__(self):
       self.stack = [] 

    def push(self, val: int) -> None:
        if len(self.stack) == 0:
            self.stack.append((val, val))
        else:
            min_value = min(self.stack[-1][1], val)
            self.stack.append((val, min_value))
        

    def pop(self) -> None:
        self.stack.pop()
        

    def top(self) -> int:
        return self.stack[-1][0]
        

    def getMin(self) -> int:
        return self.stack[-1][1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()